### PR TITLE
Only fetch requested ops for `getOps`

### PR DESCRIPTION
**Pull request by alecgibson**
# Introduction

At the moment, when calling `getOps`, `sharedb-mongo` actually fetches
all the ops from `from` to the current version.

That means that if I have a document with 1,000 versions, and I only
ask for ops 0-10, we still fetch all 1,000 ops. This is clearly
inefficient, and on documents with large numbers of ops, this fetch can
take a long time. Anecdotally, it takes ~2s to fetch ops 0-10 of
100,000 ops.

This change has an anecdotal performance increase of ~100x, now taking
~0.01s to fetch ops 0-10 of 100,000 ops.

# Background

We fetch all the ops from the current version, because we try to fetch a
"valid" chain of operations. Consider this case:

  1. I have a doc v1
  2. I submit an op v2
  3. I concurrently submit another op v2
  4. Both concurrent ops are committed to the database before the
     snapshot
  5. One of the commits successfully writes a snapshot
  6. The other snapshot is rejected, but its op is still committed
  7. I now try to get all ops. A naive `find` operation would return two
     ops for v2.

In theory, this case should actually already be [cleaned up][1], so I
can only assume that we are:

  - guarding against cases where the op cleanup fails
  - guarding against unknown inconsistencies in ops
  - guarding against stupid consumers mucking about in the db themselves
  - being very defensive

In order to therefore fetch a valid set of ops, we currently therefore
fetch the current snapshot, which has a reference stored to the op that
created it. By looking up that op, we can then check the op that
preceded _that_, and so on, forming a valid chain of ops that point to
one another.

The obvious downside of this is that we need to start at the current
snapshot and work backwards, necessitating a fetch of all the ops up to
the current version.

# Changes

This change attempts to allow us to only fetch the bare minimum number
of ops, whilst still maintaining integrity and discarding erroneous ops.

This is achieved by attempting to find the first op after our `to` op
that has a unique version. The assumption here is that any op with a
unique version is inherently valid, because it had no collisions.

For example, consider the case where I have a document with some ops:

  - v1: unique
  - v2: unique
  - v3: collision 3
  - v3: collision 3
  - v4: collision 4
  - v4: collision 4
  - v5: unique
  - v6: unique
  - ...
  - v1000: unique

If I want to fetch ops v1-v3, then we:

  - look up v4
  - find that v4 is not unique
  - look up v5
  - see that v5 is unique and therefore assumed valid
  - look backwards from v5 for a chain of valid ops

This way we don't need to fetch all the ops from v5 to the current
version.

In the case where a valid op cannot be determined, we still fall back to
fetching all ops and working backwards from the current version.

# Further work

Note that this change *only* affects the `getOps` method, notably not
touching either `getOpsToSnapshot` or `getOpsBulk`.

[1]: https://github.com/share/sharedb-mongo/blob/2d579ddb80781e987707076e932cd4e01ca066ef/index.js#L189-L193

### CODE REVIEW

**Comment by gkubisa** [comment]
The idea and implementation look good to me, however, I'd add some more tests to cover all the edge cases supported by `_getOpLink`.

**Comment by alecgibson** [comment]
@gkubisa good call - caught a bug by expanding test coverage.

**Comment by ericyhwang** [comment]
Nate's comments from the PR review meeting:
* He has a concern with an edge case around Mongo's (lack of?) guarantees about cross-collection consistency and about ops potentially getting removed by TTLs
* Flagging it and having it off by default for now is fine for now

We'll still need to take a look at the PR itself.

**Comment by alecgibson** [comment]
@ericyhwang I've hidden this feature behind a flag. Please let me know when you've had a chance to look through the code itself.

File: index.js [hunk 1]
```diff
@@ -600,6 +628,94 @@ function readOpsBulk(stream, callback) {
...
+
+  var currentVersion = currentOp && currentOp.v;
+  var previousVersion = previousOp && previousOp.v;
+  var oneBeforePreviousVersion = oneBeforePreviousOp && oneBeforePreviousOp.v;
+
+  var previousVersionWasUnique = typeof previousVersion === 'number'
+    && previousVersion !== currentVersion
+    && previousVersion !== oneBeforePreviousVersion;
+
+  if (previousVersionWasUnique) {
```
Comment (alecgibson) [inline index.js]: Turn this check into a function to shorten this function.

File: index.js [hunk 2]
```diff
@@ -600,6 +628,94 @@ function readOpsBulk(stream, callback) {
...
+    if (hasNext) {
+      cursor.next(function (error, op) {
+        if (error) return callback(error);
+        ops.push(op);
+        getFirstOpWithUniqueVersion(cursor, ops, callback);
+      });
+    } else {
+      // If there's no next op to fetch, then we now know the current version is unique so
+      // long as it doesn't match the previous version.
+      var currentVersionIsUnique = typeof currentVersion === 'number'
```
Comment (alecgibson) [inline index.js]: Could also extract this into its own function for brevity.

File: index.js (hunk 1 above)
Comment (alecgibson) [reply index.js]: Alternatively, we should pull this out into its own class, who has only these three objects (so that we don't keep the rest of the array of ops in memory that we don't need).

**Comment by alecgibson** [comment]
@ericyhwang I've addressed the comments we discussed in the pull review meeting. Namely:

  - extract complicated unique op checking into its own, tested class
  - rename the feature flag to `getOpsWithoutStrictLinking`
  - update documentation to be more detailed about behaviour with the
    flag enabled and disabled

Could you please re-review?

File: op-link-validator.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+OpLinkValidator.prototype.isAtEndOfList = function () {
+  // We ascribe a special meaning to a current op of null
+  // being that we're at the end of the list, because this
+  // is the value that the Mongo cursor will return when
+  // the cursor is exhausted
+  return this.currentOp === null;
+};
+
+OpLinkValidator.prototype._previousVersionWasUnique = function () {
+  return typeof this._previousVersion() === 'number'
```
Comment (ericyhwang) [inline op-link-validator.js]: Minor thing, might be worth calling `this._previousVersion()` once and storing it.

**Comment by ericyhwang** [comment]
[sharedb-mongo@1.0.0-beta.7](https://github.com/share/sharedb-mongo/releases/tag/v1.0.0-beta.7)
